### ISSUE TITLE
Warn about blatant recursion

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#1994]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
df() {
	df "$@" | grep -v '/snap'
}
```

#### Here's what shellcheck currently says:

Nothing

#### Here's what I wanted or expected to see:

"Did you mean `command df ...`?"

I am a dummy, I know.
I should "somehow" force limits on how much processes a user may create, I know (but I don't know precicely how is this done)

But I fork-bombed my desktop more times than I rememeber.

**Comment by koalaman** [comment]
I've done this a few times as well. There were zero occurrences in my corpus of 10k scripts, but much of it is from GitHub and this kind of error wouldn't make it past development. Due to the high potential destruction and low false positive rate when approached conservatively, a test seems worth it. The current implementation only looks at the first pipeline of the first command to catch wrapper functions like this, while avoiding false positives on any other kind of recursion.

As a hilarious side effect, ShellCheck now suggests replacing `:(){ : | :& }` with `:() { command : | command : & }`

PS: With bash and zsh you can use `FUNCNEST=10` to limit these kinds of things
